File: 10/JackTokenizer.py
import re
from utils import tokenType, keyWord
from typing import IO

SYMBOL = "{}()[].,;+-*/&|<>=~"
KEYWORDS = {kw.name.lower() for kw in keyWord}

TOKEN_PATTERNS = {
    "COMMENT": r"//.*|/\*.*?\*/|/\*\*.?\*/",
    "KEYWORD": "|".join(KEYWORDS),
    "SYMBOL": "[" + re.escape(SYMBOL) + "]",
    "IDENTIFIER": "[a-zA-Z_][a-zA-Z0-9_]*",
    "INT_CONST": r"\d+",
    "STRING_CONST": '"[^"\n]*"',
}

PATTERNS = re.compile(
    "|".join(f"(?P<{name}>{pattern})" for name, pattern in TOKEN_PATTERNS.items())
)


class JackTokenizer:
    def __init__(self, file: IO[str]):
        self.file = file.read()
        self.tokens = [match for match in PATTERNS.finditer(self.file)]
        self.current_index = -1
        self.current_token = None

    def hasMoreTokens(self) -> bool:
        return self.current_index + 1 < len(self.tokens)

    def advance(self) -> None:
        self.current_index += 1
        self.current_token = self.tokens[self.current_index]
        if self.current_token.lastgroup == "COMMENT":
            self.advance()

    def tokenType(self) -> tokenType:
        return tokenType[self.current_token.lastgroup]

    def keyWord(self) -> str:
        return self.current_token.group()

    def symbol(self) -> str:
        return self.current_token.group()

    def identifier(self) -> str:
        return self.current_token.group()

    def intVal(self) -> int:
        return int(self.current_token.group())

    def stringVal(self) -> str:
        return self.current_token.group()[1:-1]

File: 11/src/JackTokenizer.cpp
#include "JackTokenizer.h"
#include <algorithm>

JackTokenizer::JackTokenizer(std::ifstream &input)
    : file(input), currentTokenType(TokenType::UNKNOWN), currentToken("") {};

bool JackTokenizer::hasMoreTokens() {
    return !file.eof();
}

void JackTokenizer::advance() {
    currentTokenType = TokenType::UNKNOWN;
    currentToken.clear();
    char ch;

    // skip whitespace
    while (file.get(ch) && isspace(ch));
    if (file.eof()) {
        return;
    }

    // skip comments
    if (ch == '/') {
        char nextCh = file.peek();
        if (nextCh == '/') {
            // single line comment
            while (file.get(ch) && ch != '\n');
            return advance();
        }
        else if (nextCh == '*') {
            // multi-line comment
            file.get(ch);
            while (file.get(ch)) {
                if (ch == '*' && file.peek() == '/') {
                    file.get(ch);
                    return advance();
                }
            }
            throw std::runtime_error("Unterminated comment");
        }
    }

    // match string constant
    if (ch == '"') {
        currentTokenType = TokenType::STRING_CONST;
        currentToken = "";
        while (file.get(ch)) {
            if (ch == '"') {
                break;
            }
            if (ch == '\n' || file.eof()) {
                throw std::runtime_error("Unterminated string constant");
            }
            currentToken += ch;
        }
    }
    // match integer constant in the range 0..32767
    else if (isdigit(ch)) {
        currentTokenType = TokenType::INT_CONST;
        currentToken = ch;
        while (file.get(ch) && isdigit(ch)) {
            currentToken += ch;
        }
        file.unget();
        int intVal = std::stoi(currentToken);
        if (intVal < 0 || intVal > 32767) {
            throw std::runtime_error("Integer constant out of range");
        }
    }
    // match keyword or identifier
    else if (isalpha(ch) || ch == '_') {
        currentToken = ch;
        while (file.get(ch) && (isalnum(ch) || ch == '_')) {
            currentToken += ch;
        }
        file.unget();
        if (keywords.find(currentToken) != keywords.end()) {
            currentTokenType = TokenType::KEYWORD;
        }
        else {
            currentTokenType = TokenType::IDENTIFIER;
        }
    }
    // match symbols
    else if (symbols.find(ch) != symbols.end()) {
        currentTokenType = TokenType::SYMBOL;
        currentToken = ch;
    }
}

TokenType JackTokenizer::tokenType() {
    return currentTokenType;
}

std::string JackTokenizer::keyWord() {
    return currentToken;
}

char JackTokenizer::symbol() {
    return currentToken[0];
}

std::string JackTokenizer::identifier() {
    return currentToken;
}

int JackTokenizer::intVal() {
    return std::stoi(currentToken);
}

std::string JackTokenizer::stringVal() {
    return currentToken;
}